### PR TITLE
Add cancellationToken local symbol

### PR DESCRIPTION
**Pull request by jcouv**
Inside an async-iterator method that returns `IAsyncEnumerable` (but not `IAsyncEnumerator`), a variable `cancellationToken` exists. It is implemented as a field on the state machine. We initialize it with the token passed as a parameter into `GetAsyncEnumerator(CancellationToken)`.

Parameters and type parameters of this name are disallowed on such async-iterators.
There is discussion of providing a warning on parameters of type `CancellationToken` as well, but I will do that in a follow-up (see below).
I temporarily blocked usage and capture of `cancellationToken` in local functions and lambdas due to time constraints.

Fixes https://github.com/dotnet/roslyn/issues/34407

- [x] Add a binder just after the InMethodBinder
- [x] Stored the synthesized local in the method. 
- [x] Checking IsIterator early is trouble (we need to determine if the local exists before binding the body of the method, but binding the body is currently required to determine if a method is an iterator) so I'm using the return type instead. The two are equivalent (otherwise an error is produced).
- [x] AsyncRewriter hoists the synthesized local.
- [x] GetAsyncEnumerator assigns the corresponding field with value from parameter.
- [x] Made local readonly. Don't assign a debug slot to this local. Add diagnostics for conflicts.
- [x] Report an error if the `CancellationToken` type is missing (filed https://github.com/dotnet/roslyn/issues/34709)
- [x] When a local function captures `cancellationToken` the async method should use that local function's environment, rather than its own field. (temporarily blocked this scenario. Filed https://github.com/dotnet/roslyn/issues/34708 to enable)
- [x] Add a warning for parameters of type `CancellationToken` (email thread with LDM, filed https://github.com/dotnet/roslyn/issues/34710)
- [ ] test EE (see comment in DefineUserDefinedStateMachineHoistedLocal) and EnC (add a declaration for 'cancellationToken', add a write statement)

Relates to https://github.com/dotnet/roslyn/issues/24037 (async-streams umbrella)

### CODE REVIEW

File: docs/features/async-streams.md [hunk 1]
```diff
@@ -117,7 +117,8 @@ But it contains additional state:
 - a promise of a value-or-end,
 - a current yielded value of type `T`,
 - an `int` capturing the id of the thread that created it,
-- a `bool` flag indicating "dispose mode".
+- a `bool` flag indicating "dispose mode",
+- a `CancellationToken` field.
```
Comment (jaredpar) [inline docs/features/async-streams.md]: Consider noting here it's explicitly the value accessible via `cancellationToken`

File: src/Compilers/Core/Portable/SynthesizedLocalKind.cs [hunk 2]
```diff
@@ -230,6 +230,13 @@ internal enum SynthesizedLocalKind
         /// TODO: Avoid using lambdas and display classes for implementation of relaxation stubs and remove this kind.
         /// </summary>
         DelegateRelaxationReceiver = 0x101,
+
+        /// <summary>
+        /// The variable used in async-iterator methods to refer to the CancellationToken
+        /// from `GetAsyncEnumerator(CancellationToken)`.
+        /// Never actually created as a local variable. Its usages are lowered to refer to that parameter instead.
```
Comment (jaredpar) [inline src/Compilers/Core/Portable/SynthesizedLocalKind.cs]: The uses are *logically* lowered to the parameter but actually lowered to the hoisted field.

File: docs/features/async-streams.md [hunk 3]
```diff
@@ -196,6 +197,11 @@ Similarly, the kick-off method is much like those of regular iterator methods:
 }
 ```
 
+#### `cancellationToken` pseudo-local
+
+A variable named `cancellationToken` and of type `System.Threading.CancellationToken` exists in the body of an async-iterator method that returns `IAsyncEnumerable`.
+That variable is initialized with the token received by `GetAsyncEnumerator`. The compiler treats this variable as a local that is implicitly declared and is hoisted into a field.
```
Comment (jaredpar) [inline docs/features/async-streams.md]: Consider noting the local is considered `readonly`

File: src/Compilers/CSharp/Portable/Binder/CancellationTokenLocalScopeBinder.cs [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+    internal sealed class CancellationTokenLocalScopeBinder : LocalScopeBinder
+    {
+        private readonly SourceMethodSymbol _methodSymbol;
+
+        public CancellationTokenLocalScopeBinder(SourceMethodSymbol owner, Binder enclosing)
+            : base(enclosing, enclosing.Flags & ~BinderFlags.AllClearedAtExecutableCodeBoundary)
+        {
+            Debug.Assert(!enclosing.Flags.Includes(BinderFlags.InCatchFilter));
+            Debug.Assert((object)owner != null);
+            Debug.Assert(enclosing is InMethodBinder);
```
Comment (jaredpar) [inline src/Compilers/CSharp/Portable/Binder/CancellationTokenLocalScopeBinder.cs]: Why not just make the type of `enclosing` be `InMethodBinder`?

File: src/Compilers/CSharp/Portable/Binder/CancellationTokenLocalScopeBinder.cs [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+            _methodSymbol = owner;
+        }
+
+        protected override SourceLocalSymbol LookupLocal(SyntaxToken nameToken)
+            => throw ExceptionUtilities.Unreachable;
```
Comment (jaredpar) [inline src/Compilers/CSharp/Portable/Binder/CancellationTokenLocalScopeBinder.cs]: Why throw here vs. calling through `Next.`?

File: src/Compilers/CSharp/Portable/Symbols/Synthesized/SynthesizedLocal.cs [hunk 6]
```diff
@@ -79,6 +82,19 @@ public override RefKind RefKind
...
+        internal override bool IsWritableVariable
+        {
+            get
+            {
+                if (SynthesizedKind == SynthesizedLocalKind.AsyncIteratorCancellationToken)
+                {
+                    return false;
+                }
+
+                return base.IsWritableVariable;
```
Comment (jaredpar) [inline src/Compilers/CSharp/Portable/Symbols/Synthesized/SynthesizedLocal.cs]: Why is `_refKind` not considered here? Consider that `in` would not be writable.

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAsyncIteratorTests.cs [hunk 7]
```diff
@@ -5010,5 +5045,862 @@ public async System.Collections.Generic.IAsyncEnumerator<int> GetAsyncEnumerator
...
+        {
+            Console.Write(i);
+            break;
+        }
+     }
+}";
+            var comp = CreateCompilationWithAsyncIterator(new[] { source, AsyncEnumerableWithCancellation }, TestOptions.DebugExe);
+            comp.VerifyDiagnostics();
+            CompileAndVerify(comp, expectedOutput: "1 FINALLY  The operation was canceled. 1 FINALLY FINALLY2");
+        }
```
Comment (jaredpar) [inline src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAsyncIteratorTests.cs]: Consider test where you do `nameof(cancellationToken)`

File: src/Compilers/CSharp/Portable/Binder/BinderFactory.BinderFactoryVisitor.cs [hunk 8]
```diff
@@ -145,6 +145,10 @@ public override Binder VisitMethodDeclaration(MethodDeclarationSyntax methodDecl
                     {
                         method = method ?? GetMethodSymbol(methodDecl, resultBinder);
                         resultBinder = new InMethodBinder(method, resultBinder);
+                        if (method.IsAsync)
```
Comment (jaredpar) [inline src/Compilers/CSharp/Portable/Binder/BinderFactory.BinderFactoryVisitor.cs]: Expected to see `&& method.IsIterator`

File: src/Compilers/CSharp/Portable/Lowering/StateMachineRewriter/StateMachineRewriter.cs [hunk 9]
```diff
@@ -434,33 +437,46 @@ protected SynthesizedImplementationMethod GenerateIteratorGetEnumerator(MethodSy
...
+                initializeProxiedParameter(parameter);
+            }
+
+            // Initialize the cancellationToken field
+            if (method is SourceMethodSymbol sourceMethod)
+            {
+                if (sourceMethod.GetCancellationTokenLocal() is { } cancellationTokenLocal)
                 {
+                    _ = copyDest.TryGetValue(cancellationTokenLocal, out CapturedSymbolReplacement proxy);
+                    Debug.Assert(proxy is { });
```
Comment (jaredpar) [inline src/Compilers/CSharp/Portable/Lowering/StateMachineRewriter/StateMachineRewriter.cs]: `is object` is clearer here. The `is { }` would return true fro a `struct`

File: src/Compilers/CSharp/Portable/Symbols/Source/LocalFunctionSymbol.cs [hunk 10]
```diff
@@ -103,6 +109,16 @@ internal sealed class LocalFunctionSymbol : SourceMethodSymbol
 
         public Binder ParameterBinder => _binder;
 
+        internal override SynthesizedLocal GetCancellationTokenLocal()
+        {
+            if (ReferenceEquals( _lazyCancellationTokenLocal, SentinenlLocalSymbol.Instance))
```
Comment (jaredpar) [inline src/Compilers/CSharp/Portable/Symbols/Source/LocalFunctionSymbol.cs]: Nit: space after `(`

File: src/Compilers/CSharp/Portable/Binder/BinderFactory.BinderFactoryVisitor.cs (hunk 8 above)
Comment (jcouv) [reply src/Compilers/CSharp/Portable/Binder/BinderFactory.BinderFactoryVisitor.cs]: `IsIterator` only gives a correct result after we've bound the body of the method (ie. detected `yield`). 

Two options:
- We could consider adding a syntax-based detection of iterators
- The logic inside this new binder checks for `async` + `IAsyncEnumerable`. I could move that check out here instead.

File: src/Compilers/CSharp/Portable/Symbols/Source/LocalFunctionSymbol.cs [hunk 11]
```diff
@@ -179,6 +195,11 @@ private void ComputeParameters()
             if (IsAsync)
             {
                 SourceOrdinaryMethodSymbol.ReportAsyncParameterErrors(parameters, diagnostics, this.Locations[0]);
+
+                if (GetCancellationTokenLocal() is { })
```
Comment (jaredpar) [inline src/Compilers/CSharp/Portable/Symbols/Source/LocalFunctionSymbol.cs]: Think `is object` is cleaner here.

**Comment by jaredpar** [comment]
Done with review pass (iteration 2)

File: src/Compilers/CSharp/Portable/Binder/BinderFactory.BinderFactoryVisitor.cs (hunk 8 above)
Comment (jaredpar) [reply src/Compilers/CSharp/Portable/Binder/BinderFactory.BinderFactoryVisitor.cs]: This part of the code is less familiar to me so I'd tend to trust you / other judgement here.

File: src/Compilers/CSharp/Portable/Binder/CancellationTokenLocalScopeBinder.cs (hunk 5 above)
Comment (jcouv) [reply src/Compilers/CSharp/Portable/Binder/CancellationTokenLocalScopeBinder.cs]: I looked at the callers of this method and none can presently reach here. Calling to `Next` seems fine though (can't test now, but may prevent surprises in the future)

Comment (jaredpar) [reply src/Compilers/CSharp/Portable/Binder/CancellationTokenLocalScopeBinder.cs]: The one I saw that could conceivably reach here was deconstruction. Binders is a weak area for me though. It's possible I'm just mis-understanding the contract here.

---
In reply to: [271585392](https://github.com/dotnet/roslyn/pull/34692#discussion_r271585392) [](ancestors = 271585392)

**Comment by jcouv** [comment]
Closing this PR, as we'll be moving to the parameter+attribute approach instead of special variable.

**Pull request closed without merging**